managed-ledger/src/test/java/org/apache/zookeeper/MockZooKeeper.java

```diff
@@ -170,27 +172,28 @@ public synchronized void create(final String path, final byte[] data, final List
             return;
         }
 
-        executor.execute(() -> {
+        callbackExecutor.execute(() -> {
             String parent = path.substring(0, path.lastIndexOf("/"));
 
-            synchronized (MockZooKeeper.this) {
-                if (getProgrammedFailStatus()) {
-                    cb.processResult(failReturnCode.intValue(), path, ctx, null);
-                } else if (stopped) {
-                    cb.processResult(KeeperException.Code.CONNECTIONLOSS.intValue(), path, ctx, null);
-                } else if (tree.containsKey(path)) {
-                    cb.processResult(KeeperException.Code.NODEEXISTS.intValue(), path, ctx, null);
-                } else if (!parent.isEmpty() && !tree.containsKey(parent)) {
-                    cb.processResult(KeeperException.Code.NONODE.intValue(), path, ctx, null);
-                } else {
+            if (getProgrammedFailStatus()) {
+                cb.processResult(failReturnCode.intValue(), path, ctx, null);
+            } else if (stopped) {
+                cb.processResult(KeeperException.Code.CONNECTIONLOSS.intValue(), path, ctx, null);
+            } else if (tree.containsKey(path)) {
```

Comment:
But that doesn't resolve the race conditions. In multiple place we are checking the tree map and then performing some updates, and that has to be atomic (eg: the z-node version increment)

Is the deadlock caused by calling the zk callbacks with the mutex taken? If so, we could use a  lock object so that we can unlock it before triggering the callback.